### PR TITLE
[SwiftSyntax] Add accessors for source locations and test diagnostic emission

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch adds accessors on `Syntax` for `startLocation`, `endLocation`, and `sourceRange`. They have to take a file URL, because the Syntax tree is independent of a file.

Also add a test in the DiagnosticTest that gets the source locations of a file.

### CODE REVIEW

File: test/SwiftSyntax/DiagnosticTest.swift [hunk 1]
```diff
@@ -58,4 +70,36 @@ Diagnostics.test("DiagnosticEmission") {
...
+  class Visitor: SyntaxVisitor {
+    let url: URL
+    let engine: DiagnosticEngine
+    init(url: URL, engine: DiagnosticEngine) {
+      self.url = url
+      self.engine = engine
+    }
+    override func visit(_ function: FunctionDeclSyntax) {
+      let startLoc = function.identifier.startLocation(in: url)
+      engine.diagnose(.badFunction(function.identifier), location: startLoc) {
```
Comment (nkcsgexi) [inline test/SwiftSyntax/DiagnosticTest.swift]: can we issue diags to the end of the function decl to test the end location too?

**Comment by harlanhaskins** [comment]
@nkcsgexi 👍

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Review by nkcsgexi — APPROVED** [review]

**Comment by harlanhaskins** [comment]
Mutation strikes again.

@swift-ci please smoke test

**Pull request merged**